numba/core/types/functions.py

```diff
@@ -285,8 +285,43 @@ def get_call_type(self, context, args, kws):
         prefer_not = [False, True]    # new behavior preferring non-literal
         failures = _ResolutionFailures(context, self, args, kws,
                                        depth=self._depth)
+
+        from numba.core.extending_hardware import hardware_registry
+        from numba.core.extending_hardware import current_target
+        if len(context.callstack._stack) > 0:
+            target = context.callstack[0].target
+        else:
+            target = hardware_registry.get(current_target(), None)
+
+        for k, v in hardware_registry.items():
+            if v == target:
+                target_hw = v
+                break
+        else:
+            msg = ("InternalError: The hardware target for TOS is not "
+                   "registered. Given target was {}.")
+            raise ValueError(msg.format(target))
+
+        # fish out templates that are specific to the target if a target is
+        # specified
+        DEFAULT_HARDWARE = 'cpu'
+        usable = []
+        for ix, temp_cls in enumerate(self.templates):
+            # ? Need to do something about this next line
+            hw = temp_cls.metadata.get('hardware', DEFAULT_HARDWARE)
+            if hw is not None:
+                hw_clazz = hardware_registry[hw]
+                if hw_clazz in target_hw.__mro__:
+                    usable.append((temp_cls, hw_clazz, ix))
+
+        # sort templates based on hardware specificity
+        def key(x):
+            return target_hw.__mro__.index(x[1])
+        order = [x[0] for x in sorted(usable, key=key)]
```

Comment:
We can check for empty `order` here.

Reply:
done, and removed the later check if there's zero failures as I don't think it's reachable if caught here.